course5/homework10/linkedlist.hh

```diff
@@ -0,0 +1,209 @@
+#pragma once
+
+#include <memory>
+#include <iterator>
+#include <cassert>
+
+template <typename T>
+class LinkedList
+{
+private:
+    struct Node
+    {
+        Node(std::unique_ptr<Node> &&next, const T& data)
+            : next(std::move(next))
+            , data(data)
+        {}
+
+        template <typename ...Args>
+        Node(std::unique_ptr<Node> &&next, Args &&...args)
+            : next(std::move(next))
+            , data(std::forward<Args>(args)...)
+        {}
+        std::unique_ptr<Node> next;
+        T data;
+    };
+public:
+    using value_type = T;
+
+    LinkedList() = default;
+
+    // deep copy
+    LinkedList(LinkedList &rhs)
+    {
+        *this = rhs;
+    }
+
+    // move
+    LinkedList(LinkedList &&rhs)
+        : mHead(std::move(rhs.mHead))
+        , mSize(rhs.mSize)
+    {
+        rhs.mSize = 0;
+    }
+
+    ~LinkedList()
+    {
+        std::unique_ptr<Node> tmp = std::move(mHead);
+        while (tmp != nullptr)
+            tmp = std::move(tmp->next);
+    }
+
+     // deep copy
+    LinkedList& operator=(const LinkedList &rhs)
+    {
+        *this = LinkedList(std::cbegin(rhs), std::cend(rhs));
+        return *this;
+    }
+
+    // move
+    LinkedList& operator=(LinkedList &&rhs)
+    {
+        if (this == &rhs)
+            return *this;
+        mHead = std::move(rhs.mHead);
+        mSize = rhs.mSize;
+        rhs.mSize = 0;
+        return *this;
+    }
+
+    template<typename Iter>
+    LinkedList(Iter begin_iter, Iter end_iter)
+    {
+        if (begin_iter == end_iter)
+            return;
+        mHead = std::make_unique<Node>(nullptr, *begin_iter);
+        ++mSize;
+        Node *tail = mHead.get();
+        auto it = begin_iter;
+        ++it;
+        for (; it != end_iter; ++it)
+        {
+            tail->next = std::make_unique<Node>(nullptr, *it);
+            tail = tail->next.get();
+            ++mSize;
+        }
+    }
+
+    template <typename NodeType, typename ValueType>
+    class IteratorType : public std::iterator<std::forward_iterator_tag, ValueType>
+    {
+    public:
+        IteratorType(NodeType node)
+            : node(node)
+        {}
+
+        bool operator==(const IteratorType &rhs)
+        {
+            return node == rhs.node;
+        }
+
+        bool operator!=(const IteratorType &rhs)
+        {
+            return !(*this == rhs);
+        }
+
+        IteratorType& operator++()
+        {
+            node = node->next.get();
+            return *this;
+        }
+
+        IteratorType operator++(int)
+        {
+            IteratorType it = *this;
+            ++(*this);
+            return it;
+        }
+
+        ValueType& operator*()
+        {
+            return node->data;
+        }
+
+        const ValueType& operator*() const
+        {
+            return node->data;
+        }
+    private:
+        NodeType node;
```

Comment:
It's not clear that ```NoteType``` is a pointer. Using ```NodeType*``` and passing const/non const ```Node``` as template argument would be cleaner.